File: ports/elixir/lib/unicode_security/utf8.ex
defmodule UnicodeSecurity.Utf8 do
  @moduledoc """
  Strict UTF-8 codec — validator, decoder, and encoder.

  The accepted byte set is exactly the strict RFC 3629 acceptance language: it
  rejects overlong encodings, surrogate codepoints (U+D800..U+DFFF), codepoints
  beyond U+10FFFF, truncated multi-byte sequences, invalid start bytes, and
  invalid continuation bytes. Acceptance is decided by an explicit state machine
  over raw octets rather than any platform UTF-8 primitive; the accepted byte set
  is closed-form per the spec.

  Reject kinds are the six atoms enumerated by the strict reject taxonomy:
  `:overlong_encoding`, `:surrogate_codepoint`, `:codepoint_beyond_max`,
  `:truncated_sequence`, `:invalid_start_byte`, `:invalid_continuation_byte`.

  Offset convention for `first_invalid_utf8_offset/1`: the returned offset is the
  index of the byte on which the state machine transitions to reject. For
  `:overlong_encoding` the offset is the start byte of the sequence, not the last
  byte consumed. `:truncated_sequence` reports an offset equal to `byte_size`.
  """

  import Bitwise

  @type reject_kind ::
          :overlong_encoding
          | :surrogate_codepoint
          | :codepoint_beyond_max
          | :truncated_sequence
          | :invalid_start_byte
          | :invalid_continuation_byte

  @doc "Stable CamelCase sub-threat tag for a strict UTF-8 reject kind."
  @spec reject_tag(reject_kind()) :: String.t()
  def reject_tag(:overlong_encoding), do: "OverlongEncoding"
  def reject_tag(:surrogate_codepoint), do: "SurrogateCodepoint"
  def reject_tag(:codepoint_beyond_max), do: "CodepointBeyondMax"
  def reject_tag(:truncated_sequence), do: "TruncatedSequence"
  def reject_tag(:invalid_start_byte), do: "InvalidStartByte"
  def reject_tag(:invalid_continuation_byte), do: "InvalidContinuationByte"

  @doc """
  First reject offset as `{offset, kind}`, or `nil` when the input is valid
  UTF-8.
  """
  @spec first_invalid_utf8_offset(binary()) :: {non_neg_integer(), reject_kind()} | nil
  def first_invalid_utf8_offset(data) when is_binary(data) do
    walk(data, 0, 0, :expect_start, byte_size(data))
  end

  # Start state at byte i: record the sequence start.
  defp walk(<<b, rest::binary>>, i, _seq_start, :expect_start, total) do
    case step(:expect_start, b) do
      {:continue, state} -> walk(rest, i + 1, i, state, total)
      {:emit, _cp, state} -> walk(rest, i + 1, i, state, total)
      {:reject, kind} -> {i, kind}
    end
  end

  # Continuation state at byte i: keep the recorded sequence start.
  defp walk(<<b, rest::binary>>, i, seq_start, {:expect_cont, _, _, _} = state, total) do
    case step(state, b) do
      {:continue, next} -> walk(rest, i + 1, seq_start, next, total)
      {:emit, _cp, next} -> walk(rest, i + 1, seq_start, next, total)
      {:reject, :overlong_encoding} -> {seq_start, :overlong_encoding}
      {:reject, kind} -> {i, kind}
    end
  end

  defp walk(<<>>, _i, _seq_start, {:expect_cont, _, _, _}, total),
    do: {total, :truncated_sequence}

  defp walk(<<>>, _i, _seq_start, :expect_start, _total), do: nil

  @doc "Process one byte given the current decoder state."
  @spec step(term(), 0..255) ::
          {:continue, term()} | {:emit, non_neg_integer(), term()} | {:reject, reject_kind()}
  def step(:expect_start, byte) do
    n = band(byte, 0xFF)

    cond do
      n < 0x80 -> {:emit, n, :expect_start}
      n < 0xC2 -> {:reject, :invalid_start_byte}
      n < 0xE0 -> {:continue, {:expect_cont, 1, band(n, 0x1F), 0x80}}
      n < 0xF0 -> {:continue, {:expect_cont, 2, band(n, 0x0F), 0x800}}
      n < 0xF5 -> {:continue, {:expect_cont, 3, band(n, 0x07), 0x10000}}
      true -> {:reject, :invalid_start_byte}
    end
  end

  def step({:expect_cont, remaining, accum, min_cp}, byte) do
    n = band(byte, 0xFF)

    if n < 0x80 or n >= 0xC0 do
      {:reject, :invalid_continuation_byte}
    else
      nxt = bor(bsl(accum, 6), band(n, 0x3F))

      if remaining == 1 do
        cond do
          nxt < min_cp -> {:reject, :overlong_encoding}
          nxt >= 0xD800 and nxt <= 0xDFFF -> {:reject, :surrogate_codepoint}
          nxt > 0x10FFFF -> {:reject, :codepoint_beyond_max}
          true -> {:emit, nxt, :expect_start}
        end
      else
        {:continue, {:expect_cont, remaining - 1, nxt, min_cp}}
      end
    end
  end

  @doc "Whole-input validity predicate."
  @spec valid?(binary()) :: boolean()
  def valid?(data), do: first_invalid_utf8_offset(data) == nil

  @doc """
  Decode a UTF-8 byte string to a codepoint list. Semantically meaningful only
  when the input is valid UTF-8; on malformed input the walker yields the longest
  valid prefix and stops.
  """
  @spec decode_to_codepoints(binary()) :: [non_neg_integer()]
  def decode_to_codepoints(data) when is_binary(data) do
    decode(data, :expect_start, [])
  end

  defp decode(<<b, rest::binary>>, state, acc) do
    case step(state, b) do
      {:continue, next} -> decode(rest, next, acc)
      {:emit, cp, next} -> decode(rest, next, [cp | acc])
      {:reject, _kind} -> Enum.reverse(acc)
    end
  end

  defp decode(<<>>, _state, acc), do: Enum.reverse(acc)

  @doc "Encode a single codepoint as a 1–4 byte UTF-8 sequence."
  @spec encode_codepoint(non_neg_integer()) :: binary()
  def encode_codepoint(cp) when cp < 0x80, do: <<cp>>

  def encode_codepoint(cp) when cp < 0x800 do
    <<bor(0xC0, bsr(cp, 6)), bor(0x80, band(cp, 0x3F))>>
  end

  def encode_codepoint(cp) when cp < 0x10000 do
    <<bor(0xE0, bsr(cp, 12)), bor(0x80, band(bsr(cp, 6), 0x3F)), bor(0x80, band(cp, 0x3F))>>
  end

  def encode_codepoint(cp) do
    <<bor(0xF0, bsr(cp, 18)), bor(0x80, band(bsr(cp, 12), 0x3F)),
      bor(0x80, band(bsr(cp, 6), 0x3F)), bor(0x80, band(cp, 0x3F))>>
  end

  @doc "Concatenate the UTF-8 encodings of a codepoint sequence."
  @spec encode_codepoints([non_neg_integer()]) :: binary()
  def encode_codepoints(cps) do
    cps |> Enum.map(&encode_codepoint/1) |> IO.iodata_to_binary()
  end
end

File: ports/lua/src/unicode_lua/utf8.lua
-- Strict UTF-8 codec — validator and decoder.
--
-- The accepted byte set is exactly the strict RFC 3629 acceptance language: it
-- rejects overlong encodings, surrogate codepoints (U+D800..U+DFFF), codepoints
-- beyond U+10FFFF, truncated multi-byte sequences, invalid start bytes, and
-- invalid continuation bytes.  Byte inputs are 1-indexed tables of integers in
-- [0, 255]; reported offsets are 0-based, matching the shared decode fixtures.
--
-- Offset convention for `first_invalid_utf8_offset`: the returned offset is the
-- index of the byte on which the state machine transitions to reject.  For
-- `OverlongEncoding` the offset is the start byte of the sequence.

local bit = require("bit")
local strict = require("unicode_lua.strict")

local Utf8RejectKind = strict.Utf8RejectKind

local M = {}

-- Decoder state is a table:
--   { start = true }
--   { start = false, remaining = <int>, accum = <int>, min_cp = <int> }
-- A step returns one of:
--   "continue", state
--   "emit", cp, state
--   "reject", reject_kind

local function state_start()
  return { start = true }
end

-- Process one byte given the current state.
function M.utf8_decode_step(state, byte)
  local n = byte
  if state.start then
    if n < 0x80 then
      return "emit", n, state_start()
    elseif n < 0xC2 then
      return "reject", Utf8RejectKind.InvalidStartByte
    elseif n < 0xE0 then
      return "continue", { start = false, remaining = 1, accum = bit.band(n, 0x1F), min_cp = 0x80 }
    elseif n < 0xF0 then
      return "continue", { start = false, remaining = 2, accum = bit.band(n, 0x0F), min_cp = 0x800 }
    elseif n < 0xF5 then
      return "continue", { start = false, remaining = 3, accum = bit.band(n, 0x07), min_cp = 0x10000 }
    else
      return "reject", Utf8RejectKind.InvalidStartByte
    end
  else
    if n < 0x80 or n >= 0xC0 then
      return "reject", Utf8RejectKind.InvalidContinuationByte
    end
    local next_accum = bit.bor(bit.lshift(state.accum, 6), bit.band(n, 0x3F))
    if state.remaining == 1 then
      if next_accum < state.min_cp then
        return "reject", Utf8RejectKind.OverlongEncoding
      elseif next_accum >= 0xD800 and next_accum <= 0xDFFF then
        return "reject", Utf8RejectKind.SurrogateCodepoint
      elseif next_accum > 0x10FFFF then
        return "reject", Utf8RejectKind.CodepointBeyondMax
      else
        return "emit", next_accum, state_start()
      end
    else
      return "continue", { start = false, remaining = state.remaining - 1, accum = next_accum, min_cp = state.min_cp }
    end
  end
end

-- The first byte offset (0-based) at which the strict UTF-8 state machine
-- rejects, plus the reject kind; returns nil when the input is valid UTF-8.
-- `bytes` is a 1-indexed table of integers.
function M.first_invalid_utf8_offset(bytes)
  local state = state_start()
  local seq_start = 0
  local len = #bytes
  for idx = 1, len do
    local i = idx - 1
    local b = bytes[idx]
    if state.start then
      seq_start = i
    end
    local result, a, _c = M.utf8_decode_step(state, b)
    if result == "continue" then
      state = a
    elseif result == "emit" then
      state = _c
    else -- "reject"
      local kind = a
      if kind == Utf8RejectKind.OverlongEncoding then
        return seq_start, kind
      end
      return i, kind
    end
  end
  if not state.start then
    return len, Utf8RejectKind.TruncatedSequence
  end
  return nil
end

-- Whole-input validity predicate.
function M.is_valid_utf8(bytes)
  return M.first_invalid_utf8_offset(bytes) == nil
end

-- Decode a UTF-8 byte table to a codepoint table (1-indexed).  On malformed
-- input the walker yields the longest valid prefix and stops.
function M.decode_to_codepoints(bytes)
  local out = {}
  local state = state_start()
  for idx = 1, #bytes do
    local b = bytes[idx]
    local result, a, c = M.utf8_decode_step(state, b)
    if result == "continue" then
      state = a
    elseif result == "emit" then
      out[#out + 1] = a
      state = c
    else -- "reject"
      return out
    end
  end
  return out
end

-- Encode a single codepoint as a 1-4 byte UTF-8 sequence (1-indexed table).
function M.encode_codepoint(cp)
  if cp < 0x80 then
    return { cp }
  elseif cp < 0x800 then
    return {
      bit.bor(0xC0, bit.rshift(cp, 6)),
      bit.bor(0x80, bit.band(cp, 0x3F)),
    }
  elseif cp < 0x10000 then
    return {
      bit.bor(0xE0, bit.rshift(cp, 12)),
      bit.bor(0x80, bit.band(bit.rshift(cp, 6), 0x3F)),
      bit.bor(0x80, bit.band(cp, 0x3F)),
    }
  else
    return {
      bit.bor(0xF0, bit.rshift(cp, 18)),
      bit.bor(0x80, bit.band(bit.rshift(cp, 12), 0x3F)),
      bit.bor(0x80, bit.band(bit.rshift(cp, 6), 0x3F)),
      bit.bor(0x80, bit.band(cp, 0x3F)),
    }
  end
end

-- Concatenate the UTF-8 encodings of a codepoint table.
function M.encode_codepoints(cps)
  local out = {}
  for idx = 1, #cps do
    local enc = M.encode_codepoint(cps[idx])
    for j = 1, #enc do
      out[#out + 1] = enc[j]
    end
  end
  return out
end

return M
